sdk/core/core-https/test/bearerTokenAuthenticationPolicy.spec.ts

```diff
@@ -48,19 +54,66 @@ describe("BearerTokenAuthenticationPolicy", function() {
     assert.strictEqual(request.headers.get("Authorization"), `Bearer ${mockToken}`);
   });
 
-  it("refreshes access tokens when they expire", async () => {
-    const now = Date.now();
-    const refreshCred1 = new MockRefreshAzureCredential(now);
-    const refreshCred2 = new MockRefreshAzureCredential(now + DefaultTokenRefreshBufferMs);
-    const notRefreshCred1 = new MockRefreshAzureCredential(
-      now + DefaultTokenRefreshBufferMs + 5000
-    );
+  it("refreshes the token on initial request", async () => {
+    const expiresOn = Date.now() + 1000 * 60; // One minute later.
+    const credential = new MockRefreshAzureCredential(expiresOn);
+
+    const request = createPipelineRequest({ url: "https://example.com" });
+    const successResponse: PipelineResponse = {
+      headers: createHttpHeaders(),
+      request,
+      status: 200
+    };
+    const next = sinon.stub<Parameters<SendRequest>, ReturnType<SendRequest>>();
+    next.resolves(successResponse);
+
+    const policy = createBearerTokenPolicy("test-scope", credential);
+
+    await policy.sendRequest(request, next);
+    assert.strictEqual(credential.authCount, 1);
+  });
+
+  it("refreshes the token again only once it expired", async () => {
+    const expireDelayMs = 5000;
+    let tokenExpiration = Date.now() + expireDelayMs;
+    const credential = new MockRefreshAzureCredential(tokenExpiration);
+
+    const request = createPipelineRequest({ url: "https://example.com" });
+    const successResponse: PipelineResponse = {
+      headers: createHttpHeaders(),
+      request,
+      status: 200
+    };
+    const next = sinon.stub<Parameters<SendRequest>, ReturnType<SendRequest>>();
+    next.resolves(successResponse);
+
+    const policy = createBearerTokenPolicy("test-scope", credential);
+
+    // The token is cached and remains cached for a bit.
+    await policy.sendRequest(request, next);
+    await policy.sendRequest(request, next);
+    assert.strictEqual(credential.authCount, 1);
+
+    // The token will remain cached until tokenExpiration - testTokenRefreshBufferMs, so in (5000 - 1000) milliseconds.
+
+    // For safe measure, we test the token is still cached a second earlier than the refresh expectation.
+    await delay(expireDelayMs - testTokenExpiresThisEarlier - 1000);
+    await policy.sendRequest(request, next);
+    assert.strictEqual(credential.authCount, 1);
 
-    const credentialsToTest: [MockRefreshAzureCredential, number][] = [
-      [refreshCred1, 2],
-      [refreshCred2, 2],
-      [notRefreshCred1, 1]
-    ];
+    // The new token will last 5 seconds again.
+    tokenExpiration = Date.now() + expireDelayMs;
+    credential.setExpiresOnTimestamp(tokenExpiration);
+
+    // Now we wait until it expires:
+    await delay(1000);
```

Comment:
please don't use real `delay` in tests, use sinon to mock out the clock and advance time.

Reply:
ohhh that's fancy! Sounds good!

Reply:
About sinon time mock! 🤔 it only replaces setTimeout, clearTimeout, setInterval, clearInterval, which we don't use in the expiration code since we check for the raw dates and compare them. To make tests faster, it will be enough to make the delay shorter, but I will still need to wait until the delay happens (so I need delay). How does that sound?

Reply:
Disregard! I got it working